Add port: libLAS with compression support

ShinNoNoir opened the pull request:
This PR adds [`libLAS`](https://www.liblas.org/). It's a C/C++ library for reading and writing the very common LAS LiDAR format. 

The library depends on `geotiff` and optionally on `laszip` for compression support. Since compressed LAS files (`.laz`) are quite common, I've decided to make this port of `libLAS` depend on `laszip` by default.

The `geotiff` library depends on `proj`, but the existing `proj` port had some issues:
- No debug targets were exported, resulting in the debug version of `geotiff` getting linked against the release version of `proj`.
- The library was renamed from `proj4` to `proj`, but the `.cmake` configuration files did not reflect this. As a result, neither `find_package(PROJ)` or `find_package(PROJ4)` functioned properly.

ShinNoNoir committed:
[laszip] Initial port files

ShinNoNoir committed:
[geotiff] Initial port files

ShinNoNoir committed:
[geotiff] Remove doc directories

ShinNoNoir committed:
[liblas] Initial port files

ShinNoNoir committed:
[geotiff] Fix generated .cmake files

ShinNoNoir committed:
[proj] Include debug targets in CMake files

ShinNoNoir committed:
[geotiff] Hack: include find_package(PROJ4) in config cmake

This feels like a bit of a hack. If a consumer of geotiff does not
call find_package(PROJ4), CMake won't be able to resolve the PROJ4
dependency. This will result in "proj.lib" instead of its actual full
path to be added as a dependency (when using CMake to build a project),
causing an error during the linking phase. When find_package(PROJ4) is
called, CMake will add the correct paths to projd.lib and proj.lib.

ShinNoNoir committed:
[laszip] Define config vars

ras0219-msft commented on ports/liblas/boost_static_libs.patch:
```diff
@@ -0,0 +1,25 @@
+ if(WIN32)
+   if (MSVC)
+-    set(Boost_USE_STATIC_LIBS   ON)
++    set(Boost_USE_STATIC_LIBS   OFF)
```
This should be no longer needed due to our toolchain "hack"

ShinNoNoir replied:
I've just tried compiling libLAS without patching that particular line in `CMakeLists.txt` and it didn't work. Somewhere down the line the compilation process fails because it won't be able to find `pthread.h`. From `buildtrees/liblas/x86-windows-rel/CMakeFiles/CMakeError.log`:

```
Determining if the include file pthread.h exists failed with the following output:
Change Dir: D:/repos/vcpkg/buildtrees/liblas/x86-windows-rel/CMakeFiles/CMakeTmp

Run Build Command:"D:/repos/vcpkg/downloads/tools/ninja/ninja-1.7.2/ninja.exe" "cmTC_b988d"
[1/2] Building C object CMakeFiles\cmTC_b988d.dir\CheckIncludeFile.c.obj

FAILED: CMakeFiles/cmTC_b988d.dir/CheckIncludeFile.c.obj 

C:\PROGRA~2\MIB055~1\2017\PROFES~1\VC\Tools\MSVC\1411~1.255\bin\HostX64\x86\cl.exe  /nologo   /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /MDd /Zi /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\cmTC_b988d.dir\CheckIncludeFile.c.obj /FdCMakeFiles\cmTC_b988d.dir\ /FS -c CheckIncludeFile.c

CheckIncludeFile.c(1): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory

ninja: build stopped: subcommand failed.

```

mloskot commented on ports/geotiff/fix-import-prefix.patch:
```diff
@@ -0,0 +1,24 @@
+ # Compute the installation prefix relative to this file.
+ get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
+ get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
++get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
```
Isn't it a dup line?

ShinNoNoir replied:
The purpose of that extra line is to strip off an additional subdirectory of `{$CMAKE_CURRENT_LIST_FILE}`. I found that CMake's `cmExportInstallFileGenerator.cxx` basically generates _n_ copies of this line. In the case of `geotiff`, it needed to generate an additional line, but couldn't figure out how to address this properly, other than just patching the output.

mloskot commented:
@ShinNoNoir I'm one of libLAS developers. If you wish to submit some of your patches to the upstream repo, I'll be happy to accept.

ras0219-msft commented:
Thanks for the PR!

I've fixed the proj/proj4 issue by renaming proj -> proj4 and fixing up the cmake files. You should be able to simply use `find_package(PROJ4 REQUIRED)` now.

ras0219-msft commented on ports/laszip/export_targets.patch:
```diff
@@ -0,0 +1,88 @@
+From cd83bb98badbcb931d0589785411e3b08d6edab2 Mon Sep 17 00:00:00 2001
```
While I agree this patch would make laszip a better library, I don't think it's our place to do that.

Ideally, this could be proposed and accepted upstream and we'd backport the patch to the current stable version.

ShinNoNoir commented:
@mloskot The things I patch in libLAS seem to be mainly for adjusting it to the `vcpkg` environment. I'm not certain if they'd apply elsewhere outside of `vcpkg`.

I do agree with @ras0219-msft that I should see if I can send a patch to LASzip. I also see there is a new release of LASzip, so I could try if that one is compatible with libLAS.

ShinNoNoir commented:
Some modifications in the meantime for `geotiff`:
- Dropped the commits that patched instances of `PROJ4` to `PROJ`.
- Changed the `find_package(PROJ)` hack to `find_package(PROJ4)`.

I'm not sure if I have sufficient time this week to look into improving the situation regarding LASzip, so I might not continue work on this PR until next week.

ShinNoNoir commented:
A brief update on looking into `laszip` 3.1.0:
* No `.cmake` config or or target files are being exported (but this can be fixed by including a single `install(EXPORT LASZIPTargets ...)` command).
* It only exports `laszip_api.h`.
* It generates a `laszip_api.dll` file without any exports.

I'll have to look into how I can modify the CMake files to get the other header files to be exported. I'm not sure what I should do about `laszip_api.dll` though...

ras0219-msft commented:
It may be that they aren't set up to build dlls on windows. In this case, you can add a short warning and set `VCPKG_LIBRARY_LINKAGE` to `static` at the top of the portfile.

```cmake
if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
    message(STATUS "Warning: Dynamic building not supported yet. Building static.")
    set(VCPKG_LIBRARY_LINKAGE static)
endif()
```

mloskot commented:
What prevents from merging this one?

ShinNoNoir commented:
@mloskot Unfortunately, since October I've been working on different projects and hadn't had the chance at all to look further into it. This means I never managed to get libLAS working with a newer laszip.

I've also noticed that there is now a port for libgeotiff, meaning part of this original PR is now obsolete. If I get time to work on this again, I should probably start from scratch and try to create a port based on the most recent state of vpckg, its ports and the most recent versions of libLAS and laszip.

Rastaban commented:
Based on this comment from @ShinNoNoir I am going to close this PR:

>  If I get time to work on this again, I should probably start from scratch and try to create a port based on the most recent state of vpckg

Please open a new one if you find time to look into libLAS again.
